### PR TITLE
#5333: CUDA: Use scratch space appropriate to small reduction elements in Team reductions

### LINKED ISSUE #5333: Fix word_size_type inconsistencies in CUDA Team reduction, maybe HIP
In #5324 it was noted that `word_size_type` is used in the `RangePolicy` reduction case, but not the `Team` case. In looking, it seems that's not covered in the HIP implementation for any kind of reduction. Figure this out.

#5975 is a hypothesized consequence of these inconsistencies

### PR DESCRIPTION
**Pull request by PhilMiller**
Follow-on for #4156

Fixes #5333

### CODE REVIEW

**Comment by PhilMiller** [comment]
Things to consider
* [ ] HIP analogs
* [ ] `MDRangePolicy`

**Comment by PhilMiller** [comment]
On CUDA, this passes at least the reduction unit tests, including under `cuda-memcheck`, so it's not completely mistaken.

**Comment by PhilMiller** [comment]
On the other hand, it breaks compilation of `TestCuda_TeamBasic.cpp`

**Review by e10harvey — APPROVED** [review]

**Comment by PhilMiller** [comment]
#5555 addresses this for at least some scan cases

**Comment by PhilMiller** [comment]
The added test here does not fail without the later implementation change. I'm unsure why

File: core/unit_test/TestReducers.hpp
Comment (e10harvey) [inline core/unit_test/TestReducers.hpp:393]: Is there a separate test for `sizeof(Scalar) == 1`?

**Review by e10harvey — COMMENTED** [review]
Please see questions. Let me know if you want to sit down and triage this together.

**Comment by masterleinad** [comment]
You should probably use a test similar to the one added in https://github.com/kokkos/kokkos/pull/6296 to test the local memory implementation for Cuda and HIP (and small types of course).

**Comment by cz4rs** [comment]
retest this please

**Comment by cz4rs** [comment]
This actually works for me with 5dd2e4d on my desktop (Quadro P1000, cuda-12.2) :man_shrugging:. Further debugging ongoing...

File: core/src/Cuda/Kokkos_Cuda_Parallel_Team.hpp
Comment (masterleinad) [inline core/src/Cuda/Kokkos_Cuda_Parallel_Team.hpp:720]: This code path is never executed. It would be good to also have a test case that uses array reductions on small types (but that could be a follow-up).

**Comment by cz4rs** [comment]
Retest this please.

**Comment by cz4rs** [comment]
TODO:
- decide what to do with the introduced FIXMEs (some of the newly added tests are skipped for `OpenACC`, `OpenMPTarget`, `SYCL`, `point_t`, see the changes in `TestReducers.hpp` for details)
- [add more tests](https://github.com/kokkos/kokkos/pull/5334#issuecomment-1646097540), possibly after #6296 gets merged (?)
- follow-up for HIP after this PR is merged: https://github.com/kokkos/kokkos/pull/6461

I would still like to get some feedback on the PR in its current shape.

File: core/src/Cuda/Kokkos_Cuda_ReduceScan.hpp
Comment (cz4rs) [reply core/src/Cuda/Kokkos_Cuda_ReduceScan.hpp:106]: This seems to work when replacing `void` with the whole `std::conditional_t<sizeof(FunctorType::value_type)...` thing. I will check the tests and look for a nicer way to implement this.

File: core/unit_test/TestReducers.hpp
Comment (masterleinad) [inline core/unit_test/TestReducers.hpp:362]: What's the problem here? The `SYCL` implementation performs even the copies with `value_type`/`pointer_type` (and doesn't use `word_size_type`) so I wouldn't expect there to be a problem.

Comment (cz4rs) [reply core/unit_test/TestReducers.hpp]: This is now restricted to skip `reducers_struct` (which uses `array_reduce<float, 1>` etc.) and 
`int8_t` tests.

For `int8_t` we get `CUDA_ERROR_MISALIGNED_ADDRESS` error:
```
5: [ RUN      ] sycl.reducers_int8_t
5: 
5: PI CUDA ERROR:
5: 	Value:           716
5: 	Name:            CUDA_ERROR_MISALIGNED_ADDRESS
5: 	Description:     misaligned address
5: 	Function:        operator()
5: 	Source Location: /root/intel-llvm-mirror/sycl/plugins/cuda/pi_cuda.cpp:2553
5: 
5: 
5: PI CUDA ERROR:
5: 	Value:           716
5: 	Name:            CUDA_ERROR_MISALIGNED_ADDRESS
5: 	Description:     misaligned address
5: 	Function:        cuda_piextUSMFree
5: 	Source Location: /root/intel-llvm-mirror/sycl/plugins/cuda/pi_cuda.cpp:4953
5: 
5: terminate called after throwing an instance of 'sycl::_V1::runtime_error'
5:   what():  Native API failed. Native API returns: -999 (Unknown PI error) -999 (Unknown PI error)
```

`reducers_struct` simply hangs and causes a timeout. I'll be looking closer into those.

File: core/src/Cuda/Kokkos_Cuda_ReduceScan.hpp
Comment (cz4rs) [inline core/src/Cuda/Kokkos_Cuda_ReduceScan.hpp:106]: > `m_scratch_space` is only used as `pointer_type` below so it seems to make most sense to just `reinterpret_cast` the pointer to that type (the size of the memory pointed to by `pointer_type` is the same for the small value case). Ultimately, we use `word_size_type` when we want to copy between different allocations but as soon as we need the actual values (for `init`, `join`, `final`) we need to cast to `pointer_type`/`value_type` anyway. Thus, it might make sense to have `scratch_data` rather as `pointer_type` in the first place and only cast it to `word_size_type` when we copy it to other allocations.

_Originally posted by @masterleinad in https://github.com/kokkos/kokkos/pull/5334#discussion_r1353257202_

Switching to `pointer_type` fits nicely here, so I'm starting with that. I could also go ahead and change it in `Kokkos_Cuda_Parallel_Team.hpp` if that makes more sense :thinking:

File: core/unit_test/TestReducers.hpp
Comment (cz4rs) [inline core/unit_test/TestReducers.hpp:356]: TODO: narrow this down to skip the tests for specific types only (if there aren't too many of them).

**Review by masterleinad — APPROVED** [review]

File: core/unit_test/TestReducers.hpp
Comment (masterleinad) [inline core/unit_test/TestReducers.hpp:369]: Hmm.... I only see problems with `arry_reduce<float, 7>` on Intel GPUs.

Comment (cz4rs) [reply core/unit_test/TestReducers.hpp]: That's right, I have removed the spurious cases (895201b1b9c69c755b7dac9777d6c0b4fa487755). Still not sure what's the root cause though.

**Comment by masterleinad** [comment]
I pushed a fix for SYCL. For some reason, we were not using the same (proper) way of dealing with the `num_teams_done` variable as for `RangePolicy` and `MDRangePolicy`. Essentially, we were just appending to local memory used for group reduction results with a different type. That leads to unsuitable memory alignment for the `num_teams_done` variable. If preferred, I'm happy to push this commit to a follow-up pull request instead.

**Comment by cz4rs** [comment]
> I pushed a fix for SYCL. For some reason, we were not using the same (proper) way of dealing with the `num_teams_done` variable as for `RangePolicy` and `MDRangePolicy`. Essentially, we were just appending to local memory used for group reduction results with a different type. That leads to unsuitable memory alignment for the `num_teams_done` variable. If preferred, I'm happy to push this commit to a follow-up pull request instead.

Thanks for this! I think there's no need for another PR as long as we can deal with the [CI timeout](https://cloud.cees.ornl.gov/jenkins-ci/blue/organizations/jenkins/Kokkos/detail/Kokkos/14713/pipeline/49):
```
5: [ RUN      ] sycl.reducers_struct
 5/58 Test  #5: Kokkos_CoreUnitTest_SYCL1B ....................***Timeout 1500.19 sec
```

**Comment by cz4rs** [comment]
Retest this please.

**Comment by masterleinad** [comment]
> Looks like the `array_reduce<float, 7>` is still not working. I'll debug that more and restore that guard for that test.

I pushed another fix that somewhat hides the real issue but makes the test work and I expect that fix to be part of the proper solution (which should have its own pull request).

**Comment by cz4rs** [comment]
> I pushed another fix that somewhat hides the real issue but makes the test work and I expect that fix to be part of the proper solution (which should have its own pull request).

Alright! This would only leave FIXMEs for `OpenACC` and `point_t`. I will take a closer look at `point_t` part to at least narrow it down further.

**Comment by cz4rs** [comment]
@masterleinad The only `FIXME` left is for `OpenACC` (`size_t` and `double`). I look forward to getting this merged and dealing with `OpenACC` / adding more tests in a follow-up.

**Comment by masterleinad** [comment]
> @masterleinad The only `FIXME` left is for `OpenACC` (`size_t` and `double`). I look forward to getting this merged and dealing with `OpenACC` / adding more tests in a follow-up.

I have approved it already. 🙂

**Comment by cz4rs** [comment]
Retest this please.

**Comment by cz4rs** [comment]
Retest this please.

**Comment by cz4rs** [comment]
Retest this please.

**Comment by cz4rs** [comment]
Retest this please.

**Comment by cz4rs** [comment]
Rebased on top of `develop` to get fresh Jenkins logs. CI looks good, this needs another approval.

**Comment by crtrott** [comment]
Its not quite clear to me why the point_t default constructor is changed, considering that there is a `reduction_identity` overload. Maybe the team reduce does not use reduction identity??

**Pull request merged**
